dali/python/nvidia/dali/_conditionals.py

```diff
@@ -516,14 +522,34 @@ def if_stmt(self, cond, body, orelse, get_state, set_state, symbol_names, nouts)
             # We execute the merge _after_ both branches, and pretend for a moment, that it
             # can see those values produced in child scopes.
             with _cond_merge(split_predicate):
-                for new_body_val, new_orelse_val in zip(body_outputs, orelse_outputs):
+                err_msg = ("Divergent data found in different branches of `if/else` control flow"
+                           " statement. Variables in all code paths are merged into common output"
+                           " batches. The values assigned to a given variable need to have the same"
+                           " nesting structure in every code path (both `if` branches).\n"
+                           "For example, if we define a variable as a tuple in one branch, it must"
+                           " be defined as a tuple of the same length in the other branch - the"
+                           " contents of the tuples may be different. If we define a variable as"
+                           " a dictionary, the other branch must define it as a dictionary with the"
+                           " same set of keys, the values may be different.\n")
+
+                try:
+                    tree.assert_same_structure(body_outputs, orelse_outputs, check_types=True)
+                except ValueError as e:
+                    # Suppress the original exception, add DALI explanation at the beginning,
+                    # raise the full error message.
+                    raise ValueError(err_msg + str(e)) from None
```

Comment:
why not just from e?

Reply:
This gives us just one exception message, instead of something like:
```ValueError <message from Tree>

while handling this exception, another exception occurred:

ValueError <message from DALI>```

Reply:
Yeah, but it still adds the previous one, so you even get the `e` displayed twice, once as a preamble of `During handling of the above exception, another exception occurred:` and once for `The above exception was the direct cause of the following exception:` due to the `from e`. It's massive and confusing.